### PR TITLE
Improved map loading - caching

### PR DESCRIPTION
**Pull request by devo1929**

### CODE REVIEW

**Comment by devo1929** [comment]
The workflow:

- Load previously cached custom maps
- Iterate over all local map files, processing each in their own Task
- Calculate SHA of currently processed Map and attempt to locate it in the cache with SHA
- If not contained in the cache, load from .map file and add to cache
- Clean up cache, removing any old cache entries that were not found during local map iteration
- Save new state of cache
- Use current cache for UI

**Comment by devo1929** [comment]
Some benchmarks loading 1878 custom maps before and after on 4-core i7 cpu, 32gb of memory, m2 SSD:

Before: 22-24 seconds
After: 1-2 seconds

File: DXMainClient/Domain/Multiplayer/Map.cs [hunk 1]
```diff
@@ -107,72 +120,96 @@ public Map(string path, bool official)
...
         int Bases = -1;
 
+        [JsonProperty]
         string[] localSize;
 
+        [JsonProperty]
         string[] actualSize;
 
-        IniFile mapIni;
+        private IniFile CustomMapIni;
```
Comment (Metadorius) [inline DXMainClient/Domain/Multiplayer/Map.cs:190]: @Rampastring is this OK with our code style?

File: DXMainClient/Domain/Multiplayer/Map.cs [hunk 2]
```diff
@@ -346,39 +387,41 @@ public Point MapPointToMapPreviewPoint(Point mapPoint, Point previewSize, int le
...
+        /// This function provides the ability to load when needed.
+        /// </summary>
+        /// <returns></returns>
+        private IniFile GetCustomMapIniFile()
         {
-            // TODO necessary for custom map caching
-            throw new NotImplementedException();
+            if (CustomMapIni != null) return CustomMapIni;
+            
+            CustomMapIni = new IniFile {FileName = CustomMapFilePath};
```
Comment (Metadorius) [inline DXMainClient/Domain/Multiplayer/Map.cs:399]:
```suggestion
            CustomMapIni = new IniFile { FileName = CustomMapFilePath };
```

File: DXMainClient/Domain/Multiplayer/Map.cs [hunk 3]
```diff
@@ -107,72 +120,96 @@ public Map(string path, bool official)
...
         string[] localSize;
 
+        [JsonProperty]
         string[] actualSize;
 
-        IniFile mapIni;
+        private IniFile CustomMapIni;
 
+        [JsonProperty]
+        private string CustomMapFilePath;
```
Comment (Metadorius) [inline DXMainClient/Domain/Multiplayer/Map.cs:193]: Same here.

**Review by Metadorius — CHANGES_REQUESTED** [review]
The code looks good, I don't see any glaring issues. I've pointed out small bit of code style issues though.

File: DXMainClient/Domain/Multiplayer/Map.cs (hunk 1 above)
Comment (Rampastring) [reply DXMainClient/Domain/Multiplayer/Map.cs:190]: Private fields should be in lowercase.

**Comment by devo1929** [comment]
> Some benchmarks loading 1878 custom maps before and after on 4-core i7 cpu, 32gb of memory, m2 SSD:
> 
> Before: 22-24 seconds
> After: 1-2 seconds

I should also add that the _first_ time starting the client with this new code resulted in about 6-8 seconds load time, compared to the previous 22-24 seconds. That was due to the cache not having been built yet, but there was still a significant gain due to the parallel loading.

File: DXMainClient/Domain/Multiplayer/Map.cs (hunk 1 above)
Comment (devo1929) [reply DXMainClient/Domain/Multiplayer/Map.cs]: Resolved

File: DXMainClient/Domain/Multiplayer/Map.cs (hunk 2 above)
Comment (devo1929) [reply DXMainClient/Domain/Multiplayer/Map.cs]: Resolved

File: DXMainClient/Domain/Multiplayer/Map.cs (hunk 3 above)
Comment (devo1929) [reply DXMainClient/Domain/Multiplayer/Map.cs]: Resolved

**Comment by devo1929** [comment]
@Metadorius all requested changes were made.

**Review by Metadorius — APPROVED** [review]

**Pull request merged**